### PR TITLE
Create sale

### PR DESCRIPTION
**Pull request by Jack-Is-2048**

### CODE REVIEW

File: src/main/java/com/mercuryCyclists/Sales/controller/InStoreSaleController.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        return inStoreSaleService.getAllInStoreSales();
+    }
+
+    /**
+     * Gets In Store Sales by Store ID
+     */
+    @GetMapping(path = "/store/{storeId}/sales")
+    public List<InStoreSale> getStoreSales(@PathVariable("storeId") Long storeId) {
+        return inStoreSaleService.getStoreSales(storeId);
+    }
```
Comment (pedrooste) [inline src/main/java/com/mercuryCyclists/Sales/controller/InStoreSaleController.java:41]: Did we wanna update this?

File: src/main/java/com/mercuryCyclists/Sales/controller/InStoreSaleController.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    /**
+     * Add store to in store sale
+     */
+    @PostMapping(path = "/{inStoreSaleId}/store/{storeId}/add-store")
+    public InStoreSale addStoreToInStoreSale(@PathVariable("inStoreSaleId") Long inStoreSaleId,
+                                             @PathVariable("storeId") Long storeId) {
+        return inStoreSaleService.addStoreToInStoreSale(inStoreSaleId, storeId);
+    }
```
Comment (pedrooste) [inline src/main/java/com/mercuryCyclists/Sales/controller/InStoreSaleController.java]: did you end up looking if we could do this in a single req?

Comment (pedrooste) [reply src/main/java/com/mercuryCyclists/Sales/controller/InStoreSaleController.java]: ah yea I think you have done it in the /in-store-sale/store/id , this means we can get rid of this?

File: src/main/java/com/mercuryCyclists/Sales/controller/OnlineSaleController.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+    @Autowired
+    public OnlineSaleController(OnlineSaleService onlineSaleService) {
+        this.onlineSaleService = onlineSaleService;
+    }
+
+    /**
+     * Register a new Online Sale
+     */
+    @PostMapping
```
Comment (pedrooste) [inline src/main/java/com/mercuryCyclists/Sales/controller/OnlineSaleController.java:39]: No get endpoints?

Comment (pedrooste) [reply src/main/java/com/mercuryCyclists/Sales/controller/OnlineSaleController.java:39]: or updates, etc? Im assuming your still doing it

File: src/main/java/com/mercuryCyclists/Sales/entity/InStoreSale.java [hunk 4]
```diff
@@ -16,12 +16,15 @@
...
     @Column(name = "receipt_no", unique = true)
     private Integer receiptNo;
 
+    @ManyToOne(cascade = CascadeType.ALL)
+    @JoinColumn(name = "store_id", referencedColumnName = "id")
+    private Store store;
+
     public boolean validate(){
-        return receiptNo != null;
+        return super.validate() && receiptNo != null;
```
Comment (pedrooste) [inline src/main/java/com/mercuryCyclists/Sales/entity/InStoreSale.java:28]: nice i would have forgot that

File: src/main/java/com/mercuryCyclists/Sales/controller/InStoreSaleController.java (hunk 2 above)
Comment (Jack-Is-2048) [reply src/main/java/com/mercuryCyclists/Sales/controller/InStoreSaleController.java]: Yep forgot to remove this, will do

File: src/main/java/com/mercuryCyclists/Sales/service/InStoreSaleService.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    public ResponseEntity<InStoreSale> registerInStoreSale(InStoreSale inStoreSale, Long storeId) {
+        // Validate sale
+        if (!inStoreSale.validate()) {
+            throw new IllegalStateException("Invalid In Store Sale");
+        }
+
+        // Do after first validation in case quantity is null
+        if (inStoreSale.getQuantity() <= 0) {
+            throw new IllegalStateException("In Store Sale has 0 or a negative quantity");
+        }
```
Comment (pedrooste) [inline src/main/java/com/mercuryCyclists/Sales/service/InStoreSaleService.java]: Have a look at zhifas check, for a simplified v.
If you dont want to add a dependency, i would suggest putting this in the obj validate method

File: src/main/java/com/mercuryCyclists/Sales/controller/OnlineSaleController.java (hunk 3 above)
Comment (Jack-Is-2048) [reply src/main/java/com/mercuryCyclists/Sales/controller/OnlineSaleController.java:39]: https://github.com/pedrooste/MercuryCyclistsSales/blob/c64b7a284d4f924acaabec246cebce3c092adb15/src/main/java/com/mercuryCyclists/Sales/controller/OnlineSaleController.java

File: src/main/java/com/mercuryCyclists/Sales/controller/InStoreSaleController.java (hunk 1 above)
Comment (Jack-Is-2048) [reply src/main/java/com/mercuryCyclists/Sales/controller/InStoreSaleController.java:41]: https://github.com/pedrooste/MercuryCyclistsSales/pull/2#discussion_r971583429 opinion?
Won't be difficult to do, not sure if it's correct practice

File: src/main/java/com/mercuryCyclists/Sales/service/InStoreSaleService.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+            // Save the json object part
+            for (JsonObject part : partJsonObjs) {
+                saleService.updateProductPart(part);
+            }
+
+            // Save and return sale
+            inStoreSale.setStore(store.get());
+            inStoreSaleRepository.save(inStoreSale);
+            return new ResponseEntity<>(inStoreSale, HttpStatus.CREATED);
+        }
```
Comment (pedrooste) [inline src/main/java/com/mercuryCyclists/Sales/service/InStoreSaleService.java]: Nice looks good!
I would recommend splitting into 2 different methods for product and part, then chucking it in the sale service.

File: src/main/java/com/mercuryCyclists/Sales/service/InStoreSaleService.java (hunk 5 above)
Comment (Jack-Is-2048) [reply src/main/java/com/mercuryCyclists/Sales/service/InStoreSaleService.java]: Just had a look, will add it is a dependency

File: src/main/java/com/mercuryCyclists/Sales/controller/InStoreSaleController.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+    /**
+     * Updates existing in store sale based on the in store sale given
+     */
+    @PutMapping(path = "/{inStoreSaleId}")
+    public InStoreSale updateInStoreSale(@RequestBody InStoreSale inStoreSale,
+                                         @PathVariable("inStoreSaleId") Long inStoreSaleId) {
+        return inStoreSaleService.updateInStoreSale(inStoreSale, inStoreSaleId);
+    }
```
Comment (pedrooste) [inline src/main/java/com/mercuryCyclists/Sales/controller/InStoreSaleController.java]: There's probably a good argument somewhere that we shouldnt be updating pre-existing sales as it'll cause a shitshow. I don't mind either way; but we gotta make sure that we don't update the quanities

File: src/main/java/com/mercuryCyclists/Sales/service/InStoreSaleService.java [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+
+        Optional<InStoreSale> existingInSoreSale = inStoreSaleRepository.findById(inStoreSaleId);
+        if (!existingInSoreSale.isPresent()) {
+            throw new IllegalStateException(String.format("in store sale with Id %s does not exist", inStoreSaleId));
+        }
+
+        inStoreSaleRepository.save(inStoreSale);
+        return inStoreSale;
+    }
```
Comment (pedrooste) [inline src/main/java/com/mercuryCyclists/Sales/service/InStoreSaleService.java]: Yea, going on from the controller comment, I would leave this..

File: src/main/java/com/mercuryCyclists/Sales/service/OnlineSaleService.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+                Long partQuantity = partJsonObj.get("quantity").getAsLong();
+                if (partQuantity >= saleQuantity) {
+                    // Update part quantity
+                    partJsonObj.addProperty("quantity", (partQuantity - saleQuantity));
+
+                    // Save to ArrayList of JsonObjects
+                    partJsonObjs.add(partJsonObj);
+                } else {
+                    // return 303 Error
+                    return new ResponseEntity<>(onlineSale, HttpStatus.SEE_OTHER);
```
Comment (pedrooste) [inline src/main/java/com/mercuryCyclists/Sales/service/OnlineSaleService.java]: Probably want to return a message as well redirecting to the backorder endpoint. IDK how to do this though, zhifa has done it for somthign

**Review by pedrooste — CHANGES_REQUESTED** [review]
Lot to review!

File: src/main/java/com/mercuryCyclists/Sales/service/InStoreSaleService.java (hunk 5 above)
Comment (Jack-Is-2048) [reply src/main/java/com/mercuryCyclists/Sales/service/InStoreSaleService.java]: Dependency not found, just adding to validate method

File: src/main/java/com/mercuryCyclists/Sales/service/OnlineSaleService.java (hunk 9 above)
Comment (Jack-Is-2048) [reply src/main/java/com/mercuryCyclists/Sales/service/OnlineSaleService.java]: done

File: src/main/java/com/mercuryCyclists/Sales/entity/Sale.java [hunk 10]
```diff
@@ -1,5 +1,6 @@
 package com.mercuryCyclists.Sales.entity;
 
+import com.fasterxml.jackson.annotation.JsonFormat;
```
Comment (pedrooste) [inline src/main/java/com/mercuryCyclists/Sales/entity/Sale.java]: rm

**Review by pedrooste — APPROVED** [review]

**Pull request merged**